tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,371 @@
+// <copyright file="AsyncMethodDebuggerInvoker.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Datadog.Trace.Debugger.Helpers;
+using Datadog.Trace.Debugger.RateLimiting;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Instrumentation
+{
+    /// <summary>
+    /// AsyncMethodDebuggerInvoker is responsible for the managed side of async method instrumentation,
+    /// taking care of creating the state, capturing values and handling exceptions
+    /// </summary>
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class AsyncMethodDebuggerInvoker
+    {
+        internal static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(AsyncMethodDebuggerInvoker));
+
+        // We have two scenarios:
+        // 1. async method that does not participate in any other async method in i.e. has no async caller: probe placed on only one async method.
+        // 2. async method that has been called from another async method (async caller) -
+        // i.e. a probe was placed on two async methods that have a caller callee relationship (e.g. Async1 calls Async2 and we have probes on both of them)
+        // 2.1. async method that participate in a recursive call so we enter the same state machine type more than once but in a different context
+        // For case 1, it's simple because we can store all data in AsyncLocal and for each entry to the BeginMethod (through the state machine object),
+        // we can inspect the AsyncLocal to see if we already created the context or not
+        // but for cases 2 and 2.1 the AsyncLocal of the callee will drive the context from the caller
+        // so we can't deterministically know if the context is a continuation or first entry of the callee
+
+        /// <summary>
+        /// We use the AsyncLocal storage to store the debugger state of each logical invocation of the async method i.e. StateMachine.MoveNext run
+        /// Therefore, we must ensure that we have the same debugger state on reentry in same context (e.g. continuation that resulted from executing an "await" expression)
+        /// but a different debugger state when the entry is in a different context.
```

Comment:
```suggestion
        /// rather than a different debugger state when the entry is in a different context (e.g. in a recursive call into the same method).
```

Is my example correct?

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26